### PR TITLE
Fix memleak when process policies

### PR DESCRIPTION
**Pull request by JetXujing**
Fix memleak when process policies

### CODE REVIEW

**Comment by pmatilai** [comment]
This will fix a leak on early exit or case of single policy, but it should additionally be freed in the for-loop. But... I seriously doubt the policy does anything meaningful these days and the more merciful solution would be be just axe the whole thing.

Are you actually trying to use the SELinux policy stuff for something, or is this just something spotted by static analysis?

**Comment by JetXujing** [comment]
This is what I found by static analysis of the code.

**Comment by JetXujing** [comment]
> This will fix a leak on early exit or case of single policy, but it should additionally be freed in the for-loop. 

Yes, it should be freed in the for-loop

**Comment by pmatilai** [comment]
You don't need to test against NULL, and you can just assign like this, it always returns NULL for this very purpose:
`optCon = poptFreeContext(optCon)`

**Comment by JetXujing** [comment]
> You don't need to test against NULL, and you can just assign like this, it always returns NULL for this very purpose: `optCon = poptFreeContext(optCon)`

Thank you for your review comments. The poptFreeContext function has no return value, is the usage of `optCon = poptFreeContext(optCon)` reasonable?

poptFreeContext declares as follows:
`void poptFreeContext(poptContext con);`

**Comment by pmatilai** [comment]
What version of popt is that? Oh, at least the man page is wrong on that, it claims void but the actual header (on popt 1.19) is:

```
/**
 * Destroy context.
 * @param con           context
 * @return              NULL always
 */
poptContext poptFreeContext( poptContext con);
```

**Comment by pmatilai** [comment]
Okay, it returns NULL at least back to popt 1.13 and we certainly don't care about older than that.

File: build/policies.c [hunk 1]
```diff
@@ -298,6 +299,9 @@ static rpmRC processPolicies(rpmSpec spec, Package pkg, int test)
     free(name);
     free(types);
 
+    if (optCon)
```
Comment (pmatilai) [inline build/policies.c:302]: Don't bother checking for the NULL, poptFreeContext() does it anyway.

**Comment by JetXujing** [comment]
> What version of popt is that? Oh, at least the man page is wrong on that, it claims void but the actual header (on popt 1.19) is:
> 
> ```
> /**
>  * Destroy context.
>  * @param con           context
>  * @return              NULL always
>  */
> poptContext poptFreeContext( poptContext con);
> ```

The code has been updated.
The man page I looked at earlier. I reconfirmed the code and found that the return value had changed.

File: build/policies.c (hunk 1 above)
Comment (JetXujing) [reply build/policies.c]: done

**Comment by pmatilai** [comment]
Thanks for the patch.

It was also useful to spot the error in popt manual, submitted a fix for that: https://github.com/rpm-software-management/popt/pull/125
